tests/src/test/java/de/sovity/edc/e2e/DataSourceQueryParamsTest.java

```diff
@@ -0,0 +1,209 @@
+        var policyDefinition = PolicyDefinitionCreateRequest.builder()
+                .policyDefinitionId(dataOfferId)
+                .policy(UiPolicyCreateRequest.builder()
+                        .constraints(List.of())
+                        .build())
+                .build();
+
+        providerClient.uiApi().createPolicyDefinition(policyDefinition);
+    }
+
+    private void createContractDefinition() {
+        var contractDefinition = ContractDefinitionRequest.builder()
+                .contractDefinitionId(dataOfferId)
+                .accessPolicyId(dataOfferId)
+                .contractPolicyId(dataOfferId)
+                .assetSelector(List.of(UiCriterion.builder()
+                        .operandLeft(Prop.Edc.ID)
+                        .operator(UiCriterionOperator.EQ)
+                        .operandRight(UiCriterionLiteral.builder()
+                                .type(UiCriterionLiteralType.VALUE)
+                                .value(dataOfferId)
+                                .build())
+                        .build()))
+                .build();
+
+        providerClient.uiApi().createContractDefinition(contractDefinition);
+    }
+
+    private UiContractNegotiation initiateNegotiation(UiDataOffer dataOffer, UiContractOffer contractOffer) {
+        var negotiationRequest = ContractNegotiationRequest.builder()
+                .counterPartyAddress(dataOffer.getEndpoint())
+                .counterPartyParticipantId(dataOffer.getParticipantId())
+                .assetId(dataOffer.getAsset().getAssetId())
+                .contractOfferId(contractOffer.getContractOfferId())
+                .policyJsonLd(contractOffer.getPolicy().getPolicyJsonLd())
+                .build();
+
+        return consumerClient.uiApi().initiateContractNegotiation(negotiationRequest);
+    }
+
+    private UiContractNegotiation awaitNegotiationDone(String negotiationId) {
+        var negotiation = Awaitility.await().atMost(consumerConnector.timeout).until(
+                () -> consumerClient.uiApi().getContractNegotiation(negotiationId),
+                it -> it.getState().getSimplifiedState() != ContractNegotiationSimplifiedState.IN_PROGRESS
+        );
+
+        assertThat(negotiation.getState().getSimplifiedState()).isEqualTo(ContractNegotiationSimplifiedState.AGREED);
+        return negotiation;
+    }
+
+    private void initiateTransfer(UiContractNegotiation negotiation) {
+        var contractAgreementId = negotiation.getContractAgreementId();
+        var transferRequest = InitiateTransferRequest.builder()
+                .contractAgreementId(contractAgreementId)
+                .dataSinkProperties(dataAddress.getDataSinkProperties())
+                .build();
+        consumerClient.uiApi().initiateTransfer(transferRequest);
+    }
+
+    private String getProtocolEndpoint(ConnectorRemote connector) {
+        return connector.getConfig().getProtocolEndpoint().getUri().toString();
+    }
+}
```

Comment:
I am having some trouble validating this, could we have the test cases like this?

```java
var inputQueryString = "a=b"; // "a=b" encoded
var expectedDataSourceReceivedQueryString = "a=b"; // "a=b" double encoded (?)

createAsset(inputQueryString);
createPolicy();
createContractDefinitinon();
negotiate...
transfer...

assertThat(actualReceivedDataSourceQueryString).isEqualTo(expectedDataSourceReceivedQueryString )
```

Reply:
I would suggest adding a data source spy URL and to keep using the preexisting `validateDataTransferred` function that uses RestAssured to validate the return of that endpoint